src/main/java/frc/robot/subsystems/VisionControl.java

```diff
@@ -1,70 +1,241 @@
+        this.oi = oi;
+        this.vision = vision;
+        this.visionData = visionData;
+        this.chassis = chassis;
+        this.imu = imu;
+        ml = new MachineLearning(RECORD_PATH, FILE_NAME);
+        // this.shooter = shooter;
+    }
 
+    public void autoInit() {
+        chassis.initOdometry();
+        counter = 0;
+        
+        if(selector == "default") {
+            kP = 0.025; //.03
+            frontRightSpeed = p1.generated_frontRightEncoderPositions;
+            frontLeftSpeed = p1.generated_frontLeftEncoderPositions;
+            backRightSpeed = p1.generated_backRightEncoderPositions;
+            backLeftSpeed = p1.generated_backLeftEncoderPositions;
+            counterSpeed = 1.0; 
+        }
+
+        else {
+            kP = 0.0;
+            frontRightSpeed = skip;
+            frontLeftSpeed = skip;
+            backRightSpeed = skip;
+            backLeftSpeed = skip;
+            counterSpeed = 0.0;
+        }
+
+        chassis.setKP(kP);
     }
 
-    public void main() {
-        while (true) {
+    public void autoPeriodic() { // pathfind to cargo, collect it, and score it
+        imu.getvalues();
+
+        if(!RECORD_PATH) {
             update();
+            followPath();
+        }
+
+        else {
+            System.out.println("Please enable in teleop to record a new path");
+        }
+    }
 
-            visionData.Print();
-            if (oi.autoCollectButton()) { // Move the chassis so it is alligned, aim the shooter, and fire the cargo
+    public void main() {
+        boolean new_ball_data = false;
+        update();
+        imu.getvalues();
+
+        if(RECORD_PATH) {
+            record(oi.pilot.getLeftY(), oi.pilot.getRightX());
+        }
+
+        visionData.Print();
+        
+        if (oi.autoShootButton()) { // Move the chassis so it is alligned, aim the shooter, and fire the cargo
+            usingAuto = true;
+            
+            if(visionData.isHoopValid()) {
                 double error = 0;
 
                 // shooter.setAngle(desiredAngle);
                 // Shooter.setPower(desiredPower);
-                if (Math.abs(error) > chassisThreshold) {
-                    // chassis.drive(forward_speed, sidespeed, rotation);
+                if (Math.abs(error) > hoopChassisThreshold) {
+                    drive(hoop_forward_speed, hoop_sidespeed, hoop_rotation);
                 }
-                // shooter.shoot();
 
-            } else if (oi.autoShootButton()) { // go collect the nearest cargo
+                else {
+                    // shooter.shoot();
+                    
+                }
+            }
 
-                // shooter.gather();
-            } else {
-                break;
+            else {
+                System.out.println("Invalid data... aborting");
             }
         }
+
+        else if (oi.autoCollectButton()) { // go collect the nearest cargo
+            if(visionData.isBallValid()) {
```

Comment:
this code could use more structure (separate methods with clearly defined functionality)

Reply:
done